python/tvm/relay/frontend/pytorch.py

```diff
@@ -0,0 +1,1138 @@
+
+    def __init__(self, trace, input_shapes, input_types):
+
+        self._trace = trace
+        self._inputs_r = {}
+        self._params = {}
+        self._param_tensors = {}
+        self._consts = {}
+        self._ops = {}
+        self._op_inputs_r = {}
+        self._op_inputs_types = {}
+        self._op_inputs_otypes = {}
+        self._input_shapes = input_shapes if input_shapes else {}
+        self._input_types = input_types if input_types else {}
+        self._fn_param = []
+        self._relay_map = {}
+        self._nid_to_node_name = {}
+
+    def from_pytorch(self):
+        """ Construct relay nodes from trace of PyTorch graph
+
+        Currently only supports traced PyTorch format which means no control flow.
+        User must perform torch.jit.trace on a model and pass this in.
+        Future support should include support scripted models (torch.jit.script) which
+        preserves control flow.
+
+        Returns
+        -------
+        mod : tvm.relay.Module
+            The module that optimizations will be performed on.
+
+        params : dict of str to tvm.ndarray
+            Dict of converted parameters stored in tvm.ndarray format
+        """
+        # Check for missing ops
+        missing_operators = self._parse_import_prerequisites()
+
+        if missing_operators:
+            raise NotImplementedError( \
+                "The following operators are not implemented: {}".format(missing_operators))
+
+        # Translate PyTorch graph to by decorating Graph with state dict and inputs into each op
+        self._parse_inputs()
+        self._parse_params()
+        self._parse_ops()
+
+        nid = 0
+        for (op_name, operator), op_node in self._ops.items():
+            if operator == 'prim::Constant':
+                pass
+            elif operator == 'prim::ListConstruct':
+                if any(inp.debugName() in self._nid_to_node_name.keys() \
+                       for inp in op_node.inputs()):
+                    listconstr = []
+                    for i in op_node.inputs():
+                        if i.debugName() in self._nid_to_node_name.keys():
+                            listconstr.append( \
+                                self._relay_map[self._nid_to_node_name[i.debugName()]])
+                        elif i.node().kind() == 'prim::Constant':
+                            listconstr.append(int(self._consts[i.debugName()]))
+                        elif i.debugName() in self._inputs_r.keys():
+                            listconstr.append(int(self._inputs_r[i.debugName()]))
+
+                    # Unwrap for tensors
+                    if len(listconstr) == 1:
+                        listconstr = listconstr[0]
+
+                    self._relay_map[nid] = listconstr
+                    self._nid_to_node_name[op_name] = nid
+                    nid = nid+1
+            else:
+                for i in op_node.inputs():
```

Comment:
please try cleaning up this nested for and if. I don't see why double for loops are needed.